docs/cookbook/deploy-using-modal.md

```diff
@@ -4,36 +4,79 @@
 
 In this guide we will show you how you can use Modal to run programs written with Outlines on GPU in the cloud.
 
+## Requirements
+
+We recommend installing `modal` and `outlines` in a virtual environment. You can create one with:
+
+```bash
+python -m venv venv
+source venv/bin/activate
+```
+
+Then install the required packages:
+
+```bash
+pip install modal outlines
+```
+
 ## Build the image
 
-First we need to define our container image. We download the Mistral-7B-v0.1 model from HuggingFace as part of the definition of the image so it only needs to be done once (you need to provide an [access token](https://huggingface.co/settings/tokens))
+First we need to define our container image. We download the Mistral-7B-v0.1 model from HuggingFace as part of the definition of the image so it only needs to be done once (you need to provide an [access token](https://huggingface.co/settings/tokens)).
+
+Setting a token is best done by setting an environment variable `HF_TOKEN` with your token. If you do not wish to do this, we provide a commented-out line in the code to set the token directly in the code.
 
 ```python
 from modal import Image, App, gpu
+import os
 
+# This creates a modal App object. Here we set the name to "outlines-app".
+# There are other optional parameters like modal secrets, schedules, etc.
+# See the documentation here: https://modal.com/docs/reference/modal.App
 app = App(name="outlines-app")
 
+# Please set an environment variable HF_TOKEN with your Hugging Face API token.
+# The code below (the .env({...}) part) will copy the token from your local
+# environment to the container.
+# More info on Image here: https://modal.com/docs/reference/modal.Image
 outlines_image = Image.debian_slim(python_version="3.11").pip_install(
-    "outlines==0.0.37",
-    "transformers==4.38.2",
-    "datasets==2.18.0",
-    "accelerate==0.27.2",
-)
+    "outlines",
+    "transformers",
+    "datasets",
+    "accelerate",
+).env({
+    'HF_TOKEN':os.environ['HF_TOKEN']
+
+    # To set the token directly in the code, uncomment the line below and replace
+    # 'YOUR_TOKEN' with the HuggingFace access token.
+    # 'HF_TOKEN':'YOUR_TOKEN'
+})
+```
+
+## Setting the container up
+
+When running longer Modal apps, it's recommended to download your language model when the container starts, rather than when the function is called. This is to reduce latency on the first time you send a request to the container.
```

Comment:
It's because whatever is done when initializing the container is cached for future runs?

Reply:
Updated!